### PR TITLE
HIVE-24211: Replace Snapshot invalidate logic with WriteSet check for txn conflict detection

### PR DESCRIPTION
**Pull request by deniskuzZ**
…tection

### What changes were proposed in this pull request?

HIVE-24211: Replace Snapshot invalidate logic with WriteSet check for txn conflict detection.

### Why are the changes needed?

The Issue is that writeId in Snapshot has no information on partition, meaning that concurrent writes to different partitions would be seen as writes to the same non-partitioned table causing Snapshot to be obsolete.

### Does this PR introduce _any_ user-facing change?

HMS API, new method added 

### How was this patch tested?

unit tests

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/Driver.java [hunk 1]
```diff
@@ -497,38 +497,41 @@ private void runInternal(String command, boolean alreadyCompiled) throws Command
...
-
-          if (driverContext.getPlan().hasAcidResourcesInQuery()) {
-            compileInternal(context.getCmd(), true);
-            driverTxnHandler.recordValidWriteIds();
-            driverTxnHandler.setWriteIdForAcidFileSinks();
-          }
-          // Since we're reusing the compiled plan, we need to update its start time for current run
-          driverContext.getPlan().setQueryStartTime(driverContext.getQueryDisplay().getQueryStartTime());
-        }
+        } while (driverContext.isOutdatedTxn() && ++retryShapshotCnt <= maxRetrySnapshotCnt);
```
Comment (pvargacl) [inline ql/src/java/org/apache/hadoop/hive/ql/Driver.java:534]: That is an interesting change, do I understand correctly, that this says we only need to do a follow up check only, if the txn was outdated, because in the other cases the exclusive lock + the partition based writeset checking guarantees, that writeIds will be always valid at the second time? Could you add some comments here

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 2]
```diff
@@ -1408,6 +1410,43 @@ private boolean isUpdateOrDelete(Statement stmt, String conflictSQLSuffix) throw
     }
   }
 
+  public long getLatestTxnInConflict(long txnid) throws MetaException {
```
Comment (pvargacl) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1413]: Maybe add some javadoc, what does it do. I would emphasise that this method call only makes sense if the caller holds an exclusive lock, that blocks other txns to commit writes. And also, that this deliberately ignores inserts, or maybe this should be added to the DriverTxnHandler, that inserts will not invalidate the snapshot, that can cause duplicates

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 3]
```diff
@@ -3501,6 +3541,10 @@ public void addDynamicPartitions(AddDynamicPartitions rqst)
             pstmt.executeBatch();
           }
         }
+        try (PreparedStatement pstmt = dbConn.prepareStatement(TXN_COMPONENTS_DP_DELETE_QUERY)) {
```
Comment (pvargacl) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:3544]: This fix is not related to the original problem, shouldn't it be committed separately?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 4]
```diff
@@ -1408,6 +1410,43 @@ private boolean isUpdateOrDelete(Statement stmt, String conflictSQLSuffix) throw
...
 
+  public long getLatestTxnInConflict(long txnid) throws MetaException {
+    Connection dbConn = null;
+    Statement stmt = null;
+
+    try {
+      dbConn = getDbConn(Connection.TRANSACTION_READ_COMMITTED);
+      stmt = dbConn.createStatement();
+
+      String writeConflictQuery = "SELECT MAX(\"COMMITTED\".\"WS_TXNID\")" +
```
Comment (pvargacl) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1421]: This is probably a big enough change, but I was wondering if you could return all the conflicting txnIds and writeIds. In that case if the max conflicting txnId is less than the current one, you could skip two other call to the HMS for validTxnList and validWriteIdList and just remove the exceptions from the list.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 3 above)
Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:3546]: It's actually related. Operation that involve dynamic partitioning doesn't generate TXN_COMPONENTS at the locking stage, making it impossible to check for write conflict.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 4 above)
Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1421]: In most of the cases it will just decrease the performance. It could only be useful if there is a conflict with already registered in snapshot txn.

File: ql/src/java/org/apache/hadoop/hive/ql/Driver.java (hunk 1 above)
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/Driver.java:537]: added comments

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 2 above)
Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:1421]: added javadoc

**Comment by pvargacl** [comment]
LGTM +1

**Comment by deniskuzZ** [comment]
@jcamachor , could you please review this.

**Review by pvary — APPROVED** [review]
